hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/command/index/TestSecondaryIndex.scala

```diff
@@ -363,11 +341,86 @@ class TestSecondaryIndex extends HoodieSparkSqlTestBase {
+    if (HoodieSparkUtils.gteqSpark3_3) {
+      withTempDir { tmp =>
+        val tableName = generateTableName
+        val basePath = s"${tmp.getCanonicalPath}/$tableName"
+        // Step 1: Initial Insertion of Records
+        val dataGen = new HoodieTestDataGenerator()
+        val numInserts = 5
+        val hudiOpts: Map[String, String] = loadInitialBatchAndCreateSecondaryIndex(tableName, basePath, dataGen, numInserts)
+
+        // Verify initial state of secondary index
+        val initialKeys = spark.sql(s"select _row_key from $tableName limit 5").collect().map(_.getString(0))
+        validateSecondaryIndex(basePath, tableName, initialKeys)
+
+        // Step 3: Perform Update Operations on Subset of Records
+        val updateRecords = recordsToStrings(dataGen.generateUniqueUpdates(getInstantTime, 1, HoodieTestDataGenerator.TRIP_FLATTENED_SCHEMA)).asScala
+        val updateDf = spark.read.json(spark.sparkContext.parallelize(updateRecords.toSeq, 1))
+        val updateKeys = updateDf.select("_row_key").collect().map(_.getString(0))
+        val recordKeyToUpdate = updateKeys.head
+        val initialSecondaryKey = spark.sql(
+          s"SELECT key FROM hudi_metadata('$basePath') WHERE type=7 AND key LIKE '%$SECONDARY_INDEX_RECORD_KEY_SEPARATOR$recordKeyToUpdate'"
+        ).collect().map(indexKey => SecondaryIndexKeyUtils.getSecondaryKeyFromSecondaryIndexKey(indexKey.getString(0))).head
+        // update the record
+        updateDf.write.format("hudi")
+          .options(hudiOpts)
+          .option(OPERATION.key, UPSERT_OPERATION_OPT_VAL)
+          .mode(SaveMode.Append)
+          .save(basePath)
+        // Verify secondary index after updates
+        validateSecondaryIndex(basePath, tableName, updateKeys)
+
+        // Step 4: Perform Time Travel Query
+        // get the first instant on the timeline
+        val metaClient = HoodieTableMetaClient.builder()
+          .setBasePath(basePath)
+          .setConf(HoodieTestUtils.getDefaultStorageConf)
+          .build()
+        val firstInstant = metaClient.reloadActiveTimeline().filterCompletedInstants().firstInstant().get()
+        // do a time travel query with data skipping enabled
+        val readOpts = hudiOpts ++ Map(
+          HoodieMetadataConfig.ENABLE.key -> "true",
+          DataSourceReadOptions.ENABLE_DATA_SKIPPING.key -> "true"
+        )
+        val timeTravelDF = spark.read.format("hudi")
+          .options(readOpts)
+          .option("as.of.instant", firstInstant.requestedTime)
+          .load(basePath)
+        assertEquals(numInserts, timeTravelDF.count())
+        // updated record should still show in time travel view
+        assertEquals(1, timeTravelDF.where(s"_row_key = '$recordKeyToUpdate'").count())
+        // rider field (secondary key) should point to previous value
+        val secondaryKey = timeTravelDF.where(s"_row_key = '$recordKeyToUpdate'").select("rider").collect().head.getString(0)
```

Comment:
so, this test will succeed even if there is a bug where we poll sec index for time travel query right?

Reply:
can you corrupt one of the data files and ensure time travel fails. we are likely testing the reverse of it. 
i.e. if sec index pruning kicked in, the corrupted file will not be read and the query will succeed.

Reply:
> can you corrupt one of the data files and ensure time travel fails.

I did not get this part. If the data file is corrupt, then it should fail anyway irrespective of whether time travel used sec index or not. What this test checks is that we get valid previous value (as of first commit) and the screenshot in PR description shows that the secondary index pruning did not kick in at all because it does not support time travel query.

Reply:
I trust w/ your screen shot that it works today. but the purpose of adding a test case is, tomorrow if someone changes anything the test should fail. i.e. unintentionally if time travel was looking up in secondary index. 

So, to account for that. in addition to your existing assertions. can we do the following: 

- just go through ur usual test set up 
- add assertion that time travel works. 
- corrupt one of the data file which should not match if secondary index was looked up. 
- now re-execute the time travel query and assert for exception. Our theory here is that, SI lookup should not happen and hence all file slices will be read and hence the corrupt data file will be read and it will be thrown. 
here the assertion for exception is important. 
So, w/ this test, in future if someone untentionally reverts and ends up looking up in SI for a TT query, this test will fail since it may not throw exception

Reply:
Got it, i've added that scenario in the existing test. Please check the latest commit.